crates/sui-adapter/src/move_vm.rs

```diff
@@ -0,0 +1,240 @@
+    ) -> VMResult<Type> {
+        let data_store = ObjectRuntimeStore::new(remote);
+        self.runtime.loader().load_type(type_tag, &data_store)
+    }
+
+    pub(crate) fn get_type_layout<'r, S: MoveResolver>(
+        &self,
+        type_tag: &TypeTag,
+        remote: &'r S,
+    ) -> VMResult<MoveTypeLayout> {
+        let data_store = ObjectRuntimeStore::new(remote);
+        self.runtime.loader().get_type_layout(type_tag, &data_store)
+    }
+
+    pub(crate) fn get_type_tag(&self, ty: &Type) -> VMResult<TypeTag> {
+        self.runtime
+            .loader()
+            .type_to_type_tag(ty)
+            .map_err(|e| e.finish(Location::Undefined))
+    }
+
+    pub(crate) fn get_type_abilities(&self, ty: &Type) -> VMResult<AbilitySet> {
+        self.runtime
+            .loader()
+            .abilities(ty)
+            .map_err(|e| e.finish(Location::Undefined))
+    }
+
+    pub(crate) fn get_struct_type(&self, index: CachedStructIndex) -> Option<Arc<StructType>> {
+        self.runtime.loader().get_struct_type(index)
+    }
+
+    pub(crate) fn load_function<'r, S: MoveResolver>(
+        &self,
+        module_id: &ModuleId,
+        function_name: &IdentStr,
+        type_arguments: &[TypeTag],
+        remote: &'r S,
+    ) -> VMResult<LoadedFunctionInstantiation> {
+        let data_store = ObjectRuntimeStore::new(remote);
+        self.runtime.loader().load_function_public(
+            module_id,
+            function_name,
+            type_arguments,
+            &data_store,
+        )
+    }
+
+    pub(crate) fn execute_function_bypass_visibility<'r, S: MoveResolver>(
+        &self,
+        module: &ModuleId,
+        function_name: &IdentStr,
+        ty_args: Vec<TypeTag>,
+        args: Vec<impl Borrow<[u8]>>,
+        gas_meter: &mut impl GasMeter,
+        remote: &'r S,
+        extensions: &mut NativeContextExtensions,
+    ) -> VMResult<SerializedReturnValues> {
+        let mut data_store = ObjectRuntimeStore::new(remote);
+        let bypass_declared_entry_check = true;
+        self.runtime.execute_function(
+            module,
+            function_name,
+            ty_args,
+            args,
+            &mut data_store,
+            gas_meter,
+            extensions,
+            bypass_declared_entry_check,
+        )
+    }
+
+    pub(crate) fn publish_module_bundle<'r, S: MoveResolver>(
+        &self,
+        modules: Vec<Vec<u8>>,
+        sender: AccountAddress,
+        gas_meter: &mut impl GasMeter,
+        remote: &'r S,
+    ) -> VMResult<()> {
+        let mut data_store = ObjectRuntimeStore::new(remote);
+        self.runtime
+            .publish_module_bundle(modules, sender, &mut data_store, gas_meter)
+    }
+}
+
+struct ObjectRuntimeStore<'r, S: MoveResolver> {
+    remote: &'r S,
+}
+
+impl<'r, S: MoveResolver> ObjectRuntimeStore<'r, S> {
+    fn new(remote: &'r S) -> Self {
+        Self { remote }
+    }
+}
+
+impl<'r, S: MoveResolver> DataStore for ObjectRuntimeStore<'r, S> {
```

Comment:
this is another layer of indirection that we can and should remove in the VM side